packages/app-router/src/hooks/useLinking.ts

```diff
@@ -0,0 +1,66 @@
+import {useEffect} from 'react';
+import {Linking} from 'react-native';
+import urlParse from 'url-parse';
+
+import {useNavigator} from './useNavigator';
+
+/**
+ * A custom hook that manages deep linking by listening for URL changes
+ * and navigating accordingly using a navigator.
+ *
+ * @example
+ * function App() {
+ *   useLinking();
+ *
+ *   return <YourAppComponent />;
+ * }
+ *
+ * // This hook will automatically handle incoming deep links and navigate
+ * // based on the URL, such as `yourapp://path?query=param`.
+ */
+export function useLinking() {
+  const navigator = useNavigator();
+
+  /**
+   * Handles the URL passed from deep linking and navigates to the correct screen.
+   *
+   * @param {Object} params - Parameters object.
+   * @param {string | null} params.url - The URL to be processed.
+   *
+   * @example
+   * callback({ url: 'yourapp://profile?user=123' });
+   */
+  function callback({url}: {url: string | null}) {
+    if (!url) return;
+
+    try {
+      const {pathname, query} = urlParse(url);
```

Comment:
I think we need to use hostname in deep links as well. I am seeing examples of apps using deep links like `app-name://shop` and expecting it to route to path /shop. but since shop is the host based on urlParse the path doesn't match. example in docs would route path `?user=123` instead of `/profile?user=123`. not sure if we want to just check protocol or have a way to customize.

Reply:
After talking on the call the main issues surrounds how `url-parse` parses a provided URL.

```
https://www.google.com/search?id=abc

host -> www.google.com
path -> /search
```

```
myapp://app/search?id=abc

host -> app
path -> /search
```

This may cause issues depending on how the url scheme is registered and how the routes are defined. If there is no host defined in the url schemes then the path should be `/app/search` where is if the host is defined as `app` and the routes do not start with `/app` then the path would be `/search`. So we need to come up with logic that fills these parameters.